### DISCUSSION TITLE
Refactor imaffine class structure.

### DISCUSSION THREAD

**Discussion by omarocegueda** [dipy/dipy#3573] [1. Ideas & Feedback]
The objective of this issue is to discuss the pending question from #654, regarding the class structure of the imaffine module. The following class diagram describes the current class structure:
![imaffine_simple](https://cloud.githubusercontent.com/assets/1608583/8911926/e9611c74-3454-11e5-921e-29118666c08f.png)
(AffineRegistration "has a" MutualInformationMetric and it "has a" ParzenJointHistogram) This makes the code very simple and easy to understand (most users will use this specific configuration: MI metric with parzen histograms), but it sacrifices extensibility (we cannot add new metrics without touching the main AffineRegistration class, or add different density estimation algorithms without touching the metric).

We may define a more complex class structure as follows (this is just a starting point for the discussion):
![imaffine_extended](https://cloud.githubusercontent.com/assets/1608583/8911993/4c007082-3455-11e5-9ec0-4f446b71aaea.png)
Abstract classes are depicted in orange, and concrete (instantiable) classes are depicted in green. This structure is more complex but new metrics and new density estimation algorithms can be easily added. Also, here I am assuming that we will only support differentiable metrics (all metrics added must define a way to compute their gradient), because as far as we have observed this is what yields the most accurate results. Non-differentiable metrics can also be supported but it will add a lot more complexity to the structure.

**Comment by matthew-brett** [comment]
So - my persistent discomfort is with the `setup` method of `MutualInformationMetric`. 

I'm sorry Omar, I realize this is lazy of me, but could you remind me again why this would not work well?

```
hist_maker = ParzenJointHistogram(nbins)
transform = TranslationTransform3D()
metric = MutualInformationMetric(transform, static, moving,
                                 hist_maker, sampling_proportion=0.3)
```

**Comment by matthew-brett** [comment]
Ah - wait - I remember - you want to be able to call something like `setup` with a standard call signature that does not need to know the options for the particular metric.

The optimizer API requires an object with:
- `setup` - initializing the metric with a transform and images;
- `distance` - calculate distance value for some given parameters;
- `gradient` - calculate gradient of distance value for some given parameters;
- `distance_and_gradient`.

The implementation of the distance and gradient are specific to the implementations of the `histogram` created via `ParzenHistogram(nbins)` - calling specifically named methods of that object, as well as a call to the `compute_parzen_mi` function.

It feels as though there must be a way of separating these classes, where you could have your optimization API while allowing a cleaner separation of concerns, but I certainly can't see it at the moment.

**Comment by omarocegueda** [comment]
Hi @matthew-brett!,

> Ah - wait - I remember - you want to be able to call something like setup with a standard call signature that does not need to know the options for the particular metric.

Yes, the main problem is the multi-resolution strategy, which require to compare different pairs of images. We have two choices here:
1. Let `AffineRegistration` create new instances of the metric as needed (one instance for each resolution).
2. Expose a method from the metric that allows to change the images being compared.

For option 1, we need to define a unified way of creating and configuring metrics so that `AffineRegistration` can create the instances as needed. The problem here, is that different metrics will have different options, so we would need to tell `AffineRegistration` what metric class we want, and, say, a dictionary with the particular options for that metric class.

Option 2 uses the same metric instance, which can be created and configured beforehand. `AffineRegistration` doesn't need to know anything about those details, it just needs to compare two images and compute gradients w.r.t. the parameters of a specific transform. 

Regarding the separation of the histogram and the metric, I totally agree. It would be natural to just create and configure the histogram externally, and pass that histogram instance to the metric. This corresponds to the design in the diagram above, if you see the `HistogramMetric` class above receives a `histogram` (not `nbins`) in `__init__`. IMHO, this is a better design, which allows a lot of flexibility and extensibility, my only concern is from the user's perspective: asking the user to first create and configure a histogram, then create and configure a Metric with that histogram and then create and configure `AffineRegistration` with that metric seems, at first glance, a too complicated API especially if we consider that we only have one type of histogram right now.

The reason I am passing the `nbins` to the the current implementation of `MutualInformationMetric` is that I think this is going to be the most common use case: MI metric with some number of bins, and the user won't care about what kind of histogram we are using behind scene.

If we can find a way to implement the extensible "complicated" API, but also provide a way to easily select "typical" settings, so that the users don't need to know a lot of details unless they want to, that would be great!  =D
